src/rubicon/objc/api.py

```diff
@@ -893,15 +892,27 @@ class or a metaclass, an instance of :class:`ObjCClass` or
                 # it's the correct class.
                 #
                 # Refs #249.
+                #
+                # We also need to evict the cache when the instance *class*
+                # doesn't match. This can happen when an ObjCInstance subclass
+                # (like ObjCMutableDictionary) has been used - an ObjCInstance
+                # will respond to the same selectors as ObjCMutableDictionary,
+                # but it won't have the helper methods (like __setitem__) that
+                # make the wrapper useful.
+                #
+                # Refs #539.
+                #
                 if cls == ObjCInstance or isinstance(cls, ObjCInstance):
                     cached_class_name = cached.objc_class.name
-                    current_class_name = libobjc.class_getName(
-                        libobjc.object_getClass(object_ptr)
-                    ).decode("utf-8")
+                    current_class = libobjc.object_getClass(object_ptr)
+                    current_class_name = libobjc.class_getName(current_class).decode(
+                        "utf-8"
+                    )
+                    current_type = type_for_objcclass(current_class)
                     if (
-                        current_class_name != cached_class_name
-                        and not current_class_name.endswith(f"_{cached_class_name}")
-                    ):
+                        type(cached) is not current_type
+                        or current_class_name != cached_class_name
+                    ) and not current_class_name.endswith(f"_{cached_class_name}"):
```

Comment:
How about making "class_name" and the Python type part of the cache key? This seems conceptually cleaner than performing checks on the cached value and then raising a KeyError.

Reply:
The issue with using the name in the cache key is L915 - there are classes (NSWindow being one) that change their name after being initialised. If the original name is in the cache key, it won't get a cache hit when the "initialised" object is returned.